File: java-algorithm/src/main/java/com/leetcode/Problem790.java
package com.leetcode;

/**
 * 790. Domino and Tromino Tiling
 * https://leetcode.com/problems/domino-and-tromino-tiling/
 */
public class Problem790 {
    public int numTilings(int n) {
        if (n == 1) {
            return 1;
        }
        int base = (int) 1e9 + 7;
        long[] dp0 = new long[n + 1];  // end with |
        long[] dp1 = new long[n + 1];  // end with =
        long[] dp2 = new long[n + 1];  // end with 「
        long[] dp3 = new long[n + 1];  // end with 」
        dp0[1] = 1;
        dp0[2] = 1;
        dp1[2] = 1;
        dp2[2] = 1;
        dp3[2] = 1;
        for (int i = 3; i <= n; i++) {
            dp0[i] = (dp0[i - 1] + dp1[i - 1] + dp2[i - 1] + dp3[i - 1]) % base;
            dp1[i] = (dp0[i - 2] + dp1[i - 2]) % base;
            dp2[i] = (dp0[i - 2] + dp1[i - 2] + dp3[i - 1]) % base;
            dp3[i] = (dp0[i - 2] + dp1[i - 2] + dp2[i - 1]) % base;
        }
        return (int) (dp0[n] + dp1[n]) % base;
    }
}

File: python-algo/leetcode/problem_790.py
"""790. Domino and Tromino Tiling
https://leetcode.com/problems/domino-and-tromino-tiling/
"""


class Solution:
    def numTilings(self, n: int) -> int:
        dp = [[0, 0, 0] for _ in range(n + 1)]
        dp[0] = [1, 0, 0]
        base = int(1e9 + 7)
        for i in range(1, n + 1):
            dp[i][0] = dp[i - 1][0] + dp[i - 1][1] + dp[i - 1][2] + (dp[i - 2][0] if i >= 2 else 0)
            dp[i][1] = dp[i - 1][2] + (dp[i - 2][0] if i >= 2 else 0)
            dp[i][2] = dp[i - 1][1] + (dp[i - 2][0] if i >= 2 else 0)
            dp[i][0] %= base
            dp[i][1] %= base
            dp[i][2] %= base
        return dp[-1][0]